### PR TITLE
fix: test remove element from max heap failing as expected

### LINKED ISSUE #47401: Tests for `remove-an-element-from-a-max-heap` challenge passable with incorrect solution
### Describe the Issue

The issue is with the tests of the [Coding interview prep/Data structures/Remove an Element from a Max Heap] challenge.
In fact, I was able to pass the challenge by implementing a wrong remove method which only verifies the max heap property 
for the root element. Passing this challenge without a 100% correct solution will create issues for campers for the next one (the heap sort) as we usually adapt the max heap to become min heap. As a suggestion for resolving the issue is to test the max heap property at each insert/remove operation.
You can verify the issue with the code provided below. 

### Affected Page

https://www.freecodecamp.org/learn/coding-interview-prep/data-structures/remove-an-element-from-a-max-heap

### Your code

```
const isDefined = (value) => value !== undefined && value !== null;

var MaxHeap = function () {
  // Only change code below this line
  this.heap = [null];

  this.insert = (value) => {
    this.heap.push(value);

    const place = (index) => {
      const parentIndex = Math.floor(index / 2);
      if (parentIndex <= 0) return;
      if (this.heap[index] > this.heap[parentIndex]) {
        [this.heap[parentIndex], this.heap[index]] = [
          this.heap[index],
          this.heap[parentIndex],
        ];
        place(parentIndex);
      }
    };
    place(this.heap.length - 1);
  };
  //
  this.print = () => {
    return this.heap;
  };
  //
  this.remove = () => {
    const removed = this.heap[1];
    this.heap[1] = this.heap[this.heap.length - 1];
    this.heap.splice(this.heap.length - 1, 1);

    const place = (index) => {
      if (index === this.heap.length - 1) return;
      const child1Index = 2 * index;
      const child2Index = child1Index + 1;
      const child1 = this.heap[child1Index];
      const child2 = this.heap[child2Index];

      if (
        (!isDefined(child1) || this.heap[index] >= child1) &&
        (!isDefined(child2) || this.heap[index] >= child2)
      ) {
        return;
      }
      if (isDefined(child1) && this.heap[index] < child1) {
        [this.heap[index], this.heap[child1Index]] = [
          this.heap[child1Index],
          this.heap[index],
        ];
      } else if (isDefined(child2) && this.heap[index] < child2) {
        [this.heap[index], this.heap[child2Index]] = [
          this.heap[child2Index],
          this.heap[index],
        ];
      }
      place(index);
    };
    place(1);
    return removed;
  };
  // Only change code above this line
};
```

### Expected behavior

The test should fail because of the incorrect remove method.

### Screenshots

_No response_

### System

- Device: [Laptop]
- OS: [Windows 10]
- Browser: [Mozilla Firefox]
- Version: [Version 104.0 (64-bit) ]

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by KravMaguy**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #47401

I tested locally with code provided in the issue, it now fails as expected. I rewrote the code to increase the index at each function call and changed `else if isDefined(child2) && etc..` to `if `and it now passes as expected:

```
 this.remove = () => {
    const removed = this.heap[1];
    this.heap[1] = this.heap[this.heap.length - 1];
    this.heap.splice(this.heap.length - 1, 1);
    const place = (index) => {
      if (index === this.heap.length - 1) return;
      const child1Index = 2 * index;
      const child2Index = child1Index + 1;
      const child1 = this.heap[child1Index];
      const child2 = this.heap[child2Index];

      if (
        (!isDefined(child1) || this.heap[index] >= child1) &&
        (!isDefined(child2) || this.heap[index] >= child2)
      ) {
        return;
      }
      if (isDefined(child1) && this.heap[index] < child1) {
        [this.heap[index], this.heap[child1Index]] = [
          this.heap[child1Index],
          this.heap[index],
        ];
      }
      if (isDefined(child2) && this.heap[index] < child2) {
        [this.heap[index], this.heap[child2Index]] = [
          this.heap[child2Index],
          this.heap[index],
        ];
      }
      place((index += 1));
    };
    place(1);
    return removed;
  };

```

### CODE REVIEW

**Comment by gikf** [comment]
Could you share the complete code for your `MaxHeap`? I'm asking, because while pasting it into current seed code, it passes the test, but `remove` method doesn't behave correctly with the rest of code.

Changed test is not checking at any point if the element removed from maxheap is actually the greatest one.

**Comment by KravMaguy** [comment]
> Could you share the complete code for your `MaxHeap`? I'm asking, because while pasting it into current seed code, it passes the test, but `remove` method doesn't behave correctly with the rest of code.

I copied and pasted adelink's code from [issue](https://github.com/freeCodeCamp/freeCodeCamp/issues/47401) I only changed the else if and increased the parameter passed into the recursive call, this is what I used:

`const isDefined = (value) => value !== undefined && value !== null;

var MaxHeap = function () {
  // Only change code below this line
  this.heap = [null];

  this.insert = (value) => {
    this.heap.push(value);

    const place = (index) => {
      const parentIndex = Math.floor(index / 2);
      if (parentIndex <= 0) return;
      if (this.heap[index] > this.heap[parentIndex]) {
        [this.heap[parentIndex], this.heap[index]] = [
          this.heap[index],
          this.heap[parentIndex],
        ];
        place(parentIndex);
      }
    };
    place(this.heap.length - 1);
  };
  //
  this.print = () => {
    return this.heap;
  };
  //
  this.remove = () => {
    const removed = this.heap[1];
    this.heap[1] = this.heap[this.heap.length - 1];
    this.heap.splice(this.heap.length - 1, 1);

    const place = (index) => {
      if (index === this.heap.length - 1) return;
      const child1Index = 2 * index;
      const child2Index = child1Index + 1;
      const child1 = this.heap[child1Index];
      const child2 = this.heap[child2Index];

      if (
        (!isDefined(child1) || this.heap[index] >= child1) &&
        (!isDefined(child2) || this.heap[index] >= child2)
      ) {
        return;
      }
      if (isDefined(child1) && this.heap[index] < child1) {
        [this.heap[index], this.heap[child1Index]] = [
          this.heap[child1Index],
          this.heap[index],
        ];
      } 
      if (isDefined(child2) && this.heap[index] < child2) {
        [this.heap[index], this.heap[child2Index]] = [
          this.heap[child2Index],
          this.heap[index],
        ];
      }
      place(index+1);
    };
    place(1);
    return removed;
  };
  // Only change code above this line
};`
 
> Changed test is not checking at any point if the element removed from maxheap is actually the greatest one.

True let me fix that now.

**Review by KravMaguy — COMMENTED** [review]
check that removed element is the max

**Comment by gikf** [comment]
To be honest checking just single value might not be enough to ensure it's always removing the greatest element, and heap is keeping max heap property after the removals.

What do you think about augmenting it a bit:
- Instert all elements from `val`.
- Remove from `MaxHeap` elements one by one, pushing them to new array, checking after each one heap is still `isHeap`.
- When all elements are removed compare if order of removed elements is matching the order from greatest to lowest.

**Comment by Adelinked** [comment]
@gikf i agree with you this is the solution that i would follow. verifying the heap property at each step and check if the array of the removed elements is sorted at the end.

**Comment by KravMaguy** [comment]
> To be honest checking just single value might not be enough to ensure it's always removing the greatest element, and heap is keeping max heap property after the removals.
> 
> What do you think about augmenting it a bit:
> 
> * Instert all elements from `val`.
> * Remove from `MaxHeap` elements one by one, pushing them to new array, checking after each one heap is still `isHeap`.
> * When all elements are removed compare if order of removed elements is matching the order from greatest to lowest.

sure we can do it, just to clarify so you dont care anymore that the one removed is the max? (removed from the top). because you left that part out of these instructions, but it bothered you before.

**Comment by KravMaguy** [comment]
> To be honest checking just single value might not be enough to ensure it's always removing the greatest element, and heap is keeping max heap property after the removals.
> 
> What do you think about augmenting it a bit:
> 
> * Instert all elements from `val`.
> * Remove from `MaxHeap` elements one by one, pushing them to new array, checking after each one heap is still `isHeap`.
> * When all elements are removed compare if order of removed elements is matching the order from greatest to lowest.

Ive done as advised. though as a sidepoint, I believe we dont have to fill an entire array. We can set a variable to infinity and if is not less than the previous just return early. would that not work?


> To be honest checking just single value might not be enough to ensure it's always removing the greatest element, and heap is keeping max heap property after the removals.
> 
> What do you think about augmenting it a bit:
> 
> * Instert all elements from `val`.
> * Remove from `MaxHeap` elements one by one, pushing them to new array, checking after each one heap is still `isHeap`.
> * When all elements are removed compare if order of removed elements is matching the order from greatest to lowest.

Ok ive done this, see above commit, however rather than filling an array we can just return early if the current is not less than the previous yes?

**Review by KravMaguy — COMMENTED** [review]
checking if isHeap upon each removal

File: curriculum/challenges/english/10-coding-interview-prep/data-structures/remove-an-element-from-a-max-heap.md [hunk 1]
```diff
@@ -95,16 +110,23 @@ assert(
...
+  const [result, vals] = [[], [2, 15, 3, 7, 12, 7, 10, 90]];
+  vals.forEach((val) => test.insert(val));
+  for (let i = 0; i < vals.length; i++) {
+    const removed = test.remove();
+    if(removed > max) return false
+    max = removed;
+    result.push(removed);
+    test.print()[0] === null
+      ? (arr = test.print().slice(1))
+      : (arr = test.print());
```
Comment (gikf) [inline curriculum/challenges/english/10-coding-interview-prep/data-structures/remove-an-element-from-a-max-heap.md:124]:
```suggestion
    const curHeap = test.print();
    const arr = curHeap[0] === null ? curHeap.slice(1) : curHeap;
```
Suggestion for defining `arr` within loop, and slightly clearer assigning to it.

File: curriculum/challenges/english/10-coding-interview-prep/data-structures/remove-an-element-from-a-max-heap.md [hunk 2]
```diff
@@ -95,16 +110,23 @@ assert(
...
+    if (!isHeap(arr, 0, arr.length - 1)) {
+      return false;
+    }
```
Comment (gikf) [inline curriculum/challenges/english/10-coding-interview-prep/data-structures/remove-an-element-from-a-max-heap.md:127]:
```suggestion
    if (!isHeap(arr, 0, arr.length - 1)) {
      return false;
    }
```
Current place where heap is checked will skip checking heap with all elements inserted from `vals` and the last check will be done on heap that in theory should be empty.

What do you think about moving it (along with the `arr`, etc.) above of the `const removed = test.remove()`?

**Review by gikf — COMMENTED** [review]
Sure, returning early also would work.

It's looking nice, thanks for the updates. :+1: 

I have only couple little nitpicks, see the specific comments for them.

File: curriculum/challenges/english/10-coding-interview-prep/data-structures/remove-an-element-from-a-max-heap.md (hunk 2 above)
Comment (KravMaguy) [reply curriculum/challenges/english/10-coding-interview-prep/data-structures/remove-an-element-from-a-max-heap.md]: Understood, Ive done this, as well as implemented your other suggestions. Thank you for providing clear instruction.

**Review by gikf — APPROVED** [review]
Great, looks good :+1:

Checked with the two solutions from this pr and related issue, plus own one, results as expected.

**Review by raisedadead — APPROVED** [review]

**Pull request merged**